File: src/server/crdt/shopping/shopping_list_test.go
package crdt

import (
	"reflect"
	crdt "sdle-server/crdt/generic"
	"testing"
)

// === Auxiliary Functions ===

func listsEqual(a, b *ShoppingList) bool {
	if a.listID != b.listID || a.name != b.name {
		return false
	}

	if !reflect.DeepEqual(a.dotContext, b.dotContext) {
		return false
	}

	aKeys := a.items.Keys()
	bKeys := b.items.Keys()

	if len(aKeys) != len(bKeys) {
		return false
	}

	for _, key := range aKeys {
		itemA := a.items.Get(key)
		itemB := b.items.Get(key)

		if itemA.quantity.Read() != itemB.quantity.Read() || itemA.acquired.Read() != itemB.acquired.Read() {
			return false
		}
	}

	return true
}

// === Shopping List Tests ===

func TestShoppingList_NewShoppingList(t *testing.T) {
    list := NewShoppingList("replica1", "list1", "Groceries")

    if list.ReplicaID() != "replica1" {
        t.Errorf("Expected ReplicaID to be 'replica1', got %s", list.ReplicaID())
    }
    if list.Name() != "Groceries" {
        t.Errorf("Expected Name to be 'Groceries', got %s", list.Name())
    }
    if len(list.Items()) != 0 {
        t.Errorf("Expected initial items to be empty, got %v", list.Items())
    }
    if list.Context() == nil {
        t.Errorf("Expected context to be initialized, got nil")
    }
}

func TestShoppingList_PutItem(t *testing.T) {
    list1 := NewShoppingList("replica1", "list1", "Groceries")
    list2 := list1.Clone()

    delta := list1.PutItem("item1", "Milk", 5, 2)

    items := list1.Items()
    if len(items) != 1 {
        t.Errorf("Expected 1 item in the list, got %d", len(items))
    }
    if items[0].ItemID() != "item1" || items[0].Name() != "Milk" {
        t.Errorf("Expected item1 to be 'Milk', got %v", items[0])
    }
    if items[0].Quantity() != 5 {
        t.Errorf("Expected quantity to be 5, got %d", items[0].Quantity())
    }
    if items[0].Acquired() != 2 {
        t.Errorf("Expected acquired to be 2, got %d", items[0].Acquired())
    }

    list2.Join(delta)
    if !listsEqual(list1, list2) {
		t.Logf("List 1: %v", list1)
		t.Logf("List 2: %v", list2)
        t.Errorf("Expected list1 and list2 to be equal after joining with delta")
    }
}

func TestShoppingList_RemoveItem(t *testing.T) {
    list1 := NewShoppingList("replica1", "list1", "Groceries")
    list1.PutItem("item1", "Milk", 5, 2)
    list2 := list1.Clone()

    delta := list1.RemoveItem("item1")

    items := list1.Items()
    if len(items) != 0 {
        t.Errorf("Expected no items in the list after removal, got %d", len(items))
    }

    list2.Join(delta)
    if !listsEqual(list1, list2) {
		t.Logf("List 1: %v", list1)
		t.Logf("List 2: %v", list2)
		t.Logf("Delta: %v", delta)
        t.Errorf("Expected list1 and list2 to be equal after joining with delta")
    }
}

func TestShoppingList_Join(t *testing.T) {
    list1 := NewShoppingList("replica1", "list1", "Groceries")
    list1.PutItem("item1", "Milk", 5, 2)

    list2 := NewShoppingList("replica2", "list2", "Groceries")
    list2.PutItem("item2", "Bread", 3, 1)

    list1.Join(list2)

    items := list1.Items()
    if len(items) != 2 {
        t.Errorf("Expected 2 items in the list after join, got %d", len(items))
    }

    item1 := items[0]
    item2 := items[1]

    if item1.ItemID() == "item1" {
        if item1.Quantity() != 5 || item1.Acquired() != 2 {
            t.Errorf("Expected item1 to have quantity 5 and acquired 2, got quantity %d and acquired %d", item1.Quantity(), item1.Acquired())
        }
    } else if item1.ItemID() == "item2" {
        if item1.Quantity() != 3 || item1.Acquired() != 1 {
            t.Errorf("Expected item2 to have quantity 3 and acquired 1, got quantity %d and acquired %d", item1.Quantity(), item1.Acquired())
        }
    } else {
        t.Errorf("Unexpected item in the list: %v", item1)
    }

    if item2.ItemID() == "item1" {
        if item2.Quantity() != 5 || item2.Acquired() != 2 {
            t.Errorf("Expected item1 to have quantity 5 and acquired 2, got quantity %d and acquired %d", item2.Quantity(), item2.Acquired())
        }
    } else if item2.ItemID() == "item2" {
        if item2.Quantity() != 3 || item2.Acquired() != 1 {
            t.Errorf("Expected item2 to have quantity 3 and acquired 1, got quantity %d and acquired %d", item2.Quantity(), item2.Acquired())
        }
    } else {
        t.Errorf("Unexpected item in the list: %v", item2)
    }
}

func TestShoppingList_SetContext(t *testing.T) {
    list := NewShoppingList("replica1", "list1", "Groceries")
    newContext := crdt.NewDotContext()

    list.SetContext(newContext)

    if list.Context() != newContext {
        t.Errorf("Expected context to be updated, but it was not")
    }
}

func TestShoppingList_ConcurrentUpdates(t *testing.T) {
    list1 := NewShoppingList("replica1", "list1", "Groceries")
    list2 := NewShoppingList("replica2", "list1", "Groceries")

    delta1 := list1.PutItem("item1", "Milk", 5, 2)
    delta2 := list2.PutItem("item2", "Bread", 3, 1)

    list1.Join(delta2)
    list2.Join(delta1)

    items1 := list1.Items()
    items2 := list2.Items()

    if !listsEqual(list1, list2) {
        t.Errorf("Expected both lists to have the same items after concurrent updates,\n\ngot %v\n\nand %v\n", items1, items2)
    }

    if len(items1) != 2 {
        t.Errorf("Expected 2 items in the list after concurrent updates, got %d", len(items1))
    }
}



func TestShoppingList_JoinWithConflictingUpdates(t *testing.T) {
    list1 := NewShoppingList("replica1", "list1", "Groceries")
    list1.PutItem("item1", "Milk", 5, 2)

    list2 := NewShoppingList("replica2", "list1", "Groceries")
    list2.PutItem("item1", "Milk", 10, 4)

    list1.Join(list2)

    item := list1.GetItem("item1")
    if item == nil {
        t.Fatalf("Expected item1 to exist after join, got nil")
    }

    if item.Quantity() != 15 {
        t.Errorf("Expected quantity to be 15 after resolving conflict, got %d", item.Quantity())
    }
    if item.Acquired() != 6 {
        t.Errorf("Expected acquired to be 6 after resolving conflict, got %d", item.Acquired())
    }
}

func TestShoppingList_JoinWithEmptyList(t *testing.T) {
    list1 := NewShoppingList("replica1", "list1", "Groceries")
    list1.PutItem("item1", "Milk", 5, 2)

    list2 := NewShoppingList("replica2", "list2", "Groceries")

    list1.Join(list2)

    items := list1.Items()
    if len(items) != 1 {
        t.Errorf("Expected 1 item in the list after joining with empty list, got %d", len(items))
    }
    if items[0].ItemID() != "item1" {
        t.Errorf("Expected item1 to remain in the list, got %v", items[0])
    }
}

func TestShoppingList_JoinWithMultipleUpdates(t *testing.T) {
    list1 := NewShoppingList("replica1", "list1", "Groceries")
    list1.PutItem("item1", "Milk", 5, 2)

    list2 := NewShoppingList("replica2", "list2", "Groceries")
    list2.PutItem("item1", "Milk", 10, 4)
    list2.PutItem("item2", "Bread", 3, 1)

    list1.Join(list2)

    items := list1.Items()
    if len(items) != 2 {
        t.Errorf("Expected 2 items in the list after join, got %d", len(items))
    }

    item1 := list1.GetItem("item1")
    if item1.Quantity() != 15 || item1.Acquired() != 6 {
        t.Errorf("Expected item1 to have quantity 15 and acquired 6, got quantity %d and acquired %d", item1.Quantity(), item1.Acquired())
    }

    item2 := list1.GetItem("item2")
    if item2.Quantity() != 3 || item2.Acquired() != 1 {
        t.Errorf("Expected item2 to have quantity 3 and acquired 1, got quantity %d and acquired %d", item2.Quantity(), item2.Acquired())
    }
}

func TestShoppingList_Clone(t *testing.T) {
    list := NewShoppingList("replica1", "list1", "Groceries")
    list.PutItem("item1", "Milk", 5, 2)

    clone := list.Clone()

    if !listsEqual(list, clone) {
        t.Errorf("Expected cloned list to be equal to the original, got %v and %v", list, clone)
    }
}

func TestShoppingList_CloneIndependence(t *testing.T) {
	list1 := NewShoppingList("replica1", "list1", "Groceries")
	list1.PutItem("item1", "Milk", 5, 2)
	list2 := list1.Clone()

	list1.PutItem("item2", "Bread", 3, 1)

	if len(list2.Items()) != 1 {
		t.Errorf("Expected cloned list to remain unchanged after modifying original, got %d items", len(list2.Items()))
	}
}

func TestShoppingList_GetItem(t *testing.T) {
    list := NewShoppingList("replica1", "list1", "Groceries")
    list.PutItem("item1", "Milk", 5, 2)

    item := list.GetItem("item1")
    if item == nil {
        t.Fatalf("Expected to retrieve item1, got nil")
    }
    if item.ItemID() != "item1" || item.Name() != "Milk" {
        t.Errorf("Expected item1 to be 'Milk', got %v", item)
    }
    if item.Quantity() != 5 {
        t.Errorf("Expected quantity to be 5, got %d", item.Quantity())
    }
    if item.Acquired() != 2 {
        t.Errorf("Expected acquired to be 2, got %d", item.Acquired())
    }

    nonExistentItem := list.GetItem("item2")
    if nonExistentItem != nil {
        t.Errorf("Expected nil for non-existent item, got %v", nonExistentItem)
    }
}

func TestShoppingList_RemoveNonExistentItem(t *testing.T) {
    list1 := NewShoppingList("replica1", "list1", "Groceries")
    list2 := list1.Clone()

    delta := list1.RemoveItem("item1")

    if len(list1.Items()) != 0 {
        t.Errorf("Expected no items in the list after attempting to remove non-existent item, got %d", len(list1.Items()))
    }

    list2.Join(delta)
    if !listsEqual(list1, list2) {
        t.Errorf("Expected list1 and list2 to remain equal after removing non-existent item")
    }
}

func TestShoppingList_PutItemWithNegativeValues(t *testing.T) {
    list1 := NewShoppingList("replica1", "list1", "Groceries")
	list2 := list1.Clone()

    delta := list1.PutItem("item1", "Milk", -5, -2)

    items := list1.Items()
    if len(items) != 0 {
        t.Errorf("Expected 0 item in the list, got %d", len(items))
    }

	list2.Join(delta)
	if !listsEqual(list1, list2) {
		t.Errorf("Expected list1 and list2 to be equal after joining with delta")
	}
}

func TestShoppingList_RemoveItemWithOtherItems(t *testing.T) {
	list1 := NewShoppingList("replica1", "list1", "Groceries")
	list1.PutItem("item1", "Milk", 5, 2)
	list1.PutItem("item2", "Bread", 3, 1)
	list2 := list1.Clone()

	delta := list1.RemoveItem("item1")

	items := list1.Items()
	if len(items) != 1 {
		t.Errorf("Expected 1 item in the list after removal, got %d", len(items))
	}
	if items[0].ItemID() != "item2" {
		t.Errorf("Expected remaining item to be item2, got %v", items[0])
	}

	list2.Join(delta)
	if !listsEqual(list1, list2) {
		t.Errorf("Expected list1 and list2 to be equal after joining with delta")
	}
}

func TestShoppingList_RemoveAfterPutItem(t *testing.T) {
	list1 := NewShoppingList("replica1", "list1", "Groceries")
	list2 := list1.Clone()

	delta1 := list1.PutItem("item1", "Milk", 5, 2)
	delta2 := list1.RemoveItem("item1")

	if len(list1.Items()) != 0 {
		t.Errorf("Expected no items in the list after put and remove, got %d", len(list1.Items()))
	}

	list2.Join(delta1)
	list2.Join(delta2)
	if !listsEqual(list1, list2) {
		t.Errorf("Expected list1 and list2 to be equal after joining with deltas")
	}
}

func TestShoppingList_PutItemMultipleTimes(t *testing.T) {
	list1 := NewShoppingList("replica1", "list1", "Groceries")
	list1.PutItem("item1", "Milk", 5, 2)
	list2 := list1.Clone()

	delta := list1.PutItem("item1", "Milk", 3, 1)

	items := list1.Items()
	if len(items) != 1 {
		t.Errorf("Expected 1 item in the list after multiple puts, got %d", len(items))
	}
	if items[0].Quantity() != 8 {
		t.Errorf("Expected quantity to be 8 after multiple puts, got %d", items[0].Quantity())
	}
	if items[0].Acquired() != 3 {
		t.Errorf("Expected acquired to be 3 after multiple puts, got %d", items[0].Acquired())
	}

	list2.Join(delta)
	if !listsEqual(list1, list2) {
		t.Errorf("Expected list1 and list2 to be equal after joining with deltas")
	}
}

func TestShoppingList_EmptyList(t *testing.T) {
    list := NewShoppingList("replica1", "list1", "Groceries")

    if len(list.Items()) != 0 {
        t.Errorf("Expected empty list to have no items, got %d", len(list.Items()))
    }

    item := list.GetItem("item1")
    if item != nil {
        t.Errorf("Expected nil for non-existent item in empty list, got %v", item)
    }
}

File: src/client/types/crdt/shopping/shopping-list.test.ts
import ShoppingList from './shopping-list';
import DotContext from '../generic/dot-context';
import util from 'util';

function listsEqual(a: ShoppingList, b: ShoppingList): boolean {
    if (a.getListId() !== b.getListId() || a.getName() !== b.getName()) {
        return false;
    }

    if (JSON.stringify(a.getContext()) !== JSON.stringify(b.getContext())) {
        return false;
    }

    const aKeys = Array.from(a['items'].keys());
    const bKeys = Array.from(b['items'].keys());

    if (aKeys.length !== bKeys.length) {
        return false;
    }

    aKeys.forEach((key) => {
        const itemA = a['items'].get(key);
        const itemB = b['items'].get(key);

        if (
            itemA.getQuantity() !== itemB.getQuantity() ||
            itemA.getAcquired() !== itemB.getAcquired()
        ) {
            return false;
        }
    });

    return true;
}

describe('ShoppingList', () => {
    it('should create a new ShoppingList', () => {
        const list = new ShoppingList('replica1', 'list1', 'Groceries');

        expect(list.getReplicaId()).toBe('replica1');
        expect(list.getName()).toBe('Groceries');
        expect(list.getItems().length).toBe(0);
        expect(list.getContext()).not.toBeNull();
    });

    it('should add an item to the ShoppingList', () => {
        const list1 = new ShoppingList('replica1', 'list1', 'Groceries');
        const list2 = list1.clone();

        const delta = list1.putItem('item1', 5, 2, 'Milk');

        const items = list1.getItems();
        expect(items.length).toBe(1);
        expect(items[0].getItemId()).toBe('item1');
        expect(items[0].getName()).toBe('Milk');
        expect(items[0].getQuantity()).toBe(5);
        expect(items[0].getAcquired()).toBe(2);

        list2.join(delta);
        expect(listsEqual(list1, list2)).toBe(true);
    });

    it('should remove an item from the ShoppingList', () => {
        const list1 = new ShoppingList('replica1', 'list1', 'Groceries');
        list1.putItem('item1', 5, 2, 'Milk');
        const list2 = list1.clone();

        const delta = list1.removeItem('item1');

        const items = list1.getItems();
        expect(items.length).toBe(0);

        list2.join(delta);
        expect(listsEqual(list1, list2)).toBe(true);
    });

    it('should join two ShoppingLists', () => {
        const list1 = new ShoppingList('replica1', 'list1', 'Groceries');
        list1.putItem('item1', 5, 2, 'Milk');

        const list2 = new ShoppingList('replica2', 'list2', 'Groceries');
        list2.putItem('item2', 3, 1, 'Bread');

        list1.join(list2);

        const items = list1.getItems();
        expect(items.length).toBe(2);

        const item1 = items.find((item) => item.getItemId() === 'item1');
        const item2 = items.find((item) => item.getItemId() === 'item2');

        expect(item1?.getQuantity()).toBe(5);
        expect(item1?.getAcquired()).toBe(2);

        expect(item2?.getQuantity()).toBe(3);
        expect(item2?.getAcquired()).toBe(1);
    });

    it('should set a new context', () => {
        const list = new ShoppingList('replica1', 'list1', 'Groceries');
        const newContext = new DotContext();

        list.setContext(newContext);

        expect(list.getContext()).toBe(newContext);
    });

    it('should handle concurrent updates', () => {
        const list1 = new ShoppingList('replica1', 'list1', 'Groceries');
        const list2 = new ShoppingList('replica2', 'list1', 'Groceries');

        const delta1 = list1.putItem('item1', 5, 2, 'Milk');
        const delta2 = list2.putItem('item2', 3, 1, 'Bread');

        list1.join(delta2);
        list2.join(delta1);

        expect(listsEqual(list1, list2)).toBe(true);

        const items = list1.getItems();
        expect(items.length).toBe(2);
    });

    it('should handle conflicting updates', () => {
        const list1 = new ShoppingList('replica1', 'list1', 'Groceries');
        list1.putItem('item1', 5, 2, 'Milk');

        const list2 = new ShoppingList('replica2', 'list1', 'Groceries');
        list2.putItem('item1', 10, 4, 'Milk');

        list1.join(list2);

        const item = list1.getItem('item1');
        expect(item).not.toBeUndefined();
        expect(item?.getQuantity()).toBe(15);
        expect(item?.getAcquired()).toBe(6);
    });

    it('should clone a ShoppingList', () => {
        const list = new ShoppingList('replica1', 'list1', 'Groceries');
        list.putItem('item1', 5, 2, 'Milk');

        const clone = list.clone();

        expect(listsEqual(list, clone)).toBe(true);

        list.putItem('item2', 3, 1, 'Bread');
        expect(clone.getItems().length).toBe(1);
    });

    it('should serialize and deserialize an empty ShoppingList', () => {
        const list = new ShoppingList('replica1', 'list1', 'Groceries');

        const proto = list.toProto();
        const converted = ShoppingList.fromProto(proto);

        expect(listsEqual(list, converted)).toBe(true);
    });

    it('should serialize and deserialize a ShoppingList with items', () => {
        const list = new ShoppingList('replica1', 'list1', 'Groceries');
        list.putItem('item1', 5, 2, 'Milk');
        list.putItem('item2', 3, 1, 'Bread');

        const proto = list.toProto();
        const converted = ShoppingList.fromProto(proto);

        expect(listsEqual(list, converted)).toBe(true);
    });

    it('should join with an empty ShoppingList', () => {
        const list1 = new ShoppingList('replica1', 'list1', 'Groceries');
        list1.putItem('item1', 5, 2, 'Milk');

        const list2 = new ShoppingList('replica2', 'list2', 'Groceries');

        list1.join(list2);

        const items = list1.getItems();
        expect(items.length).toBe(1);
        expect(items[0].getItemId()).toBe('item1');
    });

    it('should join with multiple updates', () => {
        const list1 = new ShoppingList('replica1', 'list1', 'Groceries');
        list1.putItem('item1', 5, 2, 'Milk');

        const list2 = new ShoppingList('replica2', 'list2', 'Groceries');
        list2.putItem('item1', 10, 4, 'Milk');
        list2.putItem('item2', 3, 1, 'Bread');

        list1.join(list2);

        const items = list1.getItems();
        expect(items.length).toBe(2);

        const item1 = list1.getItem('item1');
        expect(item1?.getQuantity()).toBe(15);
        expect(item1?.getAcquired()).toBe(6);

        const item2 = list1.getItem('item2');
        expect(item2?.getQuantity()).toBe(3);
        expect(item2?.getAcquired()).toBe(1);
    });

    it('should maintain clone independence', () => {
        const list1 = new ShoppingList('replica1', 'list1', 'Groceries');
        list1.putItem('item1', 5, 2, 'Milk');
        const list2 = list1.clone();

        list1.putItem('item2', 3, 1, 'Bread');

        expect(list2.getItems().length).toBe(1);
    });

    it('should get an item from the ShoppingList', () => {
        const list = new ShoppingList('replica1', 'list1', 'Groceries');
        list.putItem('item1', 5, 2, 'Milk');

        const item = list.getItem('item1');
        expect(item).not.toBeUndefined();
        expect(item?.getItemId()).toBe('item1');
        expect(item?.getName()).toBe('Milk');
        expect(item?.getQuantity()).toBe(5);
        expect(item?.getAcquired()).toBe(2);

        const nonExistentItem = list.getItem('item2');
        expect(nonExistentItem).toBeUndefined();
    });

    it('should handle removing a non-existent item', () => {
        const list1 = new ShoppingList('replica1', 'list1', 'Groceries');
        const list2 = list1.clone();

        const delta = list1.removeItem('item1');

        expect(list1.getItems().length).toBe(0);

        list2.join(delta);
        expect(listsEqual(list1, list2)).toBe(true);
    });

    it('should handle putting an item with negative values', () => {
        const list1 = new ShoppingList('replica1', 'list1', 'Groceries');
        const list2 = list1.clone();

        const delta = list1.putItem('item1', -5, -2, 'Milk');

        expect(list1.getItems().length).toBe(0);

        list2.join(delta);
        expect(listsEqual(list1, list2)).toBe(true);
    });

    it('should remove an item while keeping other items intact', () => {
        const list1 = new ShoppingList('replica1', 'list1', 'Groceries');
        list1.putItem('item1', 5, 2, 'Milk');
        list1.putItem('item2', 3, 1, 'Bread');
        const list2 = list1.clone();

        const delta = list1.removeItem('item2');

        const items = list1.getItems();
        expect(items.length).toBe(1);
        expect(items[0].getItemId()).toBe('item1');

        list2.join(delta);
        expect(list2.getItems().length).toBe(1);
        expect(listsEqual(list1, list2)).toBe(true);
    });

    it('should handle removing an item after putting it', () => {
        const list1 = new ShoppingList('replica1', 'list1', 'Groceries');
        const list2 = list1.clone();

        const delta1 = list1.putItem('item1', 5, 2, 'Milk');
        const delta2 = list1.removeItem('item1');

        expect(list1.getItems().length).toBe(0);

        list2.join(delta1);
        list2.join(delta2);

        expect(listsEqual(list1, list2)).toBe(true);
    });

    it('should handle putting an item multiple times', () => {
        const list1 = new ShoppingList('replica1', 'list1', 'Groceries');
        list1.putItem('item1', 5, 2, 'Milk');
        const list2 = list1.clone();

        const delta = list1.putItem('item1', 3, 1, 'Milk');

        const items = list1.getItems();
        expect(items.length).toBe(1);
        expect(items[0].getQuantity()).toBe(8);
        expect(items[0].getAcquired()).toBe(3);

        list2.join(delta);
        expect(listsEqual(list1, list2)).toBe(true);
    });

    it('should serialize and deserialize a ShoppingList with multiple items', () => {
        const list = new ShoppingList('replica1', 'list1', 'Groceries');
        list.putItem('item1', 5, 2, 'Milk');
        list.putItem('item2', 3, 1, 'Bread');

        const proto = list.toProto();
        const converted = ShoppingList.fromProto(proto);

        expect(listsEqual(list, converted)).toBe(true);
    });

    it('should serialize and deserialize a ShoppingList with an empty item', () => {
        const list = new ShoppingList('replica1', 'list1', 'Groceries');
        list.putItem('item1', 0, 0, 'Milk');

        const proto = list.toProto();
        const converted = ShoppingList.fromProto(proto);

        expect(listsEqual(list, converted)).toBe(true);
    });

    it('should serialize and deserialize a ShoppingList with conflicting updates', () => {
        const list1 = new ShoppingList('replica1', 'list1', 'Groceries');
        list1.putItem('item1', 5, 2, 'Milk');

        const list2 = new ShoppingList('replica2', 'list1', 'Groceries');
        list2.putItem('item1', 10, 4, 'Milk');

        list1.join(list2);

        const proto = list1.toProto();
        const converted = ShoppingList.fromProto(proto);

        expect(listsEqual(list1, converted)).toBe(true);
    });

    it('should serialize and deserialize a ShoppingList after removing an item', () => {
        const list = new ShoppingList('replica1', 'list1', 'Groceries');
        list.putItem('item1', 5, 2, 'Milk');
        list.putItem('item2', 3, 1, 'Bread');

        list.removeItem('item1');

        const proto = list.toProto();
        const converted = ShoppingList.fromProto(proto);

        expect(listsEqual(list, converted)).toBe(true);
    });

    it('should serialize and deserialize a complex ShoppingList scenario', () => {
        const list = new ShoppingList('replica1', 'list1', 'Groceries');
        list.putItem('item1', 5, 2, 'Milk');
        list.putItem('item2', 3, 1, 'Bread');
        list.putItem('item3', 12, 6, 'Eggs');

        list.removeItem('item2');
        list.putItem('item4', 2, 0, 'Butter');

        const proto = list.toProto();
        const converted = ShoppingList.fromProto(proto);

        expect(listsEqual(list, converted)).toBe(true);
    });
});